docs/advanced/checkpointing.md

```diff
@@ -0,0 +1,108 @@
+# Checkpointing
+
+To perform stateful operations reliably and with consistent outputs, Quix Streams applications need to periodically save their state stores to disk and commit processed offsets to Kafka.
+
+We call this process a “checkpointing”.
+
+The goal of checkpointing is to ensure that applications can recover after failures and reprocess records from Kafka producing the same results as if the failure never happened.
+
+Currently, Quix Streams provides *At-Least-Once* processing guarantees, which means that each incoming message will be processed, but it may happen multiple times and generate duplicated outputs.
+
+## Under the Hood
+
+- When `Application` starts processing, it initializes the `Checkpoint` object with a fixed lifetime interval in seconds according to the `commit_interval` setting (every 5 seconds by default).
+- The `Checkpoint` object is responsible for keeping track of processed Kafka offsets and pending state transactions.
+- After the message is successfully processed, its offset is marked as processed in the current checkpoint.
+- When checkpoint commits, it will:
+    1. *Produce changelog messages for every pending state update to the changelog topics.*
+
+        Changelog topics are used to back up state stores in case of failures, and they contain records for every updated state key.
+        
+    2. *Flush the Kafka Producer and verify the delivery of every outgoing message both to output and changelog topics.*
+        
+        If some messages are not delivered, the checkpoint will fail.
+        
+        In this case, the messages need to be reprocessed again starting from the previous committed offset.
```

Comment:
Not sure if you intended this to still be here thanks to your "example failures" section. 

If so, maybe just link to that section with "see X for details"